feat(datasource): add Gitlab Package support (#10909)
renovatebot/renovate: lib/datasource/gitlab-packages/readme.md

[Gitlab Generic Packages Repository](https://docs.gitlab.com/ee/user/packages/generic_packages/) supports looking up package versions from all type of packages registry supported by Gitlab and can be used in combination with [regex managers](https://docs.renovatebot.com/modules/manager/regex/) to keep dependencies up-to-date which are not specifically supported by Renovate.

> To specify which specific repository should be queried when looking up a package, the `registryUrl` has to be set like this: `https://gitlab.com/user/project`. As an example, `https://gitlab.com/gitlab-org/ci-cd/package-stage/feature-testing/new-packages-list` would look for packages in the generic packages repository of the
> `gitlab-org/ci-cd/package-stage/feature-testing/new-packages-list` project.

Reviewer: ```suggestion
To specify which specific repository should be queried when looking up a package, the `registryUrl` has to be set like this: `https://gitlab.com/user/project`.
As an example, `https://gitlab.com/gitlab-org/ci-cd/package-stage/feature-testing/new-packages-list` would look for packages in the generic packages repository of the `gitlab-org/ci-cd/package-stage/feature-testing/new-packages-list` project.
```

We use the "one sentence per line" style in the docs.

**Usage Example**

> A real world example for this specific datasource would be maintaining packages versions in a config file. This can be achieved by configuring a generic regex manager in `renovate.json` for files named `versions.ini`:

Reviewer: ```suggestion
A real world example for this specific datasource would be maintaining packages versions in a config file.
This can be achieved by configuring a generic regex manager in `renovate.json` for files named `versions.ini`:
```

```json
{
  "regexManagers": [
    {
      "fileMatch": ["^versions.ini$"],
      "matchStrings": [
        "# renovate: datasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?( registryUrl=(?<registryUrl>.*?))?\\s.*?_VERSION=(?<currentValue>.*)\\s"
      ],
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}"
    }
  ]
}
```

Now you may use comments in your `versions.ini` files to automatically update dependencies, which could look like this:

```ini
# renovate: datasource=gitlab-packages depName=@gitlab-org/nk-js versioning=semver registryUrl=https://gitlab.com/gitlab-org/ci-cd/package-stage/feature-testing/new-packages-list
NKJS_VERSION=3.4.0

```

> By default, the `gitlab-packages` use the `loose` versioning scheme.

Reviewer: ```suggestion
By default, `gitlab-packages` uses the `loose` versioning scheme.
```

I think this is a bit better.

> [Gitlab Generic Packages Repository](https://docs.gitlab.com/ee/user/packages/generic_packages/) supports looking up package versions from all type of packages registry supported by Gitlab and can be used in combination with [regex managers](https://docs.renovatebot.com/modules/manager/regex/) to keep dependencies up-to-date which are not specifically supported by Renovate.

Reviewer: ```suggestion
[GitLab Generic Packages Repository](https://docs.gitlab.com/ee/user/packages/generic_packages/) supports looking up package versions from all types of packages registry supported by GitLab and can be used in combination with [regex managers](https://docs.renovatebot.com/modules/manager/regex/) to keep dependencies up-to-date which are not specifically supported by Renovate.
```

I think `types` should be plural?

Capitalize nouns like GitLab, GitHub, Docker, etc. 😉

> [GitLab Generic Packages Repository](https://docs.gitlab.com/ee/user/packages/generic_packages/) supports looking up package versions from all types of packages registry supported by GitLab and can be used in combination with [regex managers](https://docs.renovatebot.com/modules/manager/regex/) to keep dependencies up-to-date which are not specifically supported by Renovate.

Reviewer 2: Add a note that this requires GitLab >= 13.5

Author: The [package api](https://docs.gitlab.com/ee/api/packages.html) was introduced in Gitlab 11.8. It's a simple API that support listing packages from all packages registries supported by Gitlab (As of now: Generic, Composer, Conan, Debian, Go Proxy, Helm, Maven, npm, nuget, PyPi and Ruby Gems. Gitlab 13.5 introduced the support for the generic package registry, but this code will also function for all other type of packages.

For performance reason, It's better to use it with Gitlab >=12.9, because of the added "package_name" argument, that allow to only get info about the package we are looking (fuzzy search by lookupName).

To specify which specific repository should be queried when looking up a package, the `registryUrl` has to be set like this: `https://gitlab.com/user/project`.
As an example, `https://gitlab.com/gitlab-org/ci-cd/package-stage/feature-testing/new-packages-list` would look for packages in the generic packages repository of the `gitlab-org/ci-cd/package-stage/feature-testing/new-packages-list` project.

**Usage Example**

> A real world example for this specific datasource would be maintaining packages versions in a config file.

Reviewer: ```suggestion
A real-world example for this specific datasource would be maintaining packages versions in a config file.
```

I'm not sure which is best here:

- `real world`
- `real-world`

Can a native English speaker decide which is best? 😉

Author: Mmh, I'm not native, but according to [Qora](https://www.quora.com/Should-real-world-be-hyphenated#:~:text=The%20simplest%20answer%20you%20can,%2C%20pronouns%20or%20noun%20phrases.%5D), in this case it's real-world.

[GitLab Packages API](https://docs.gitlab.com/ee/api/packages.html) supports looking up package versions from [all types of packages registry supported by GitLab](https://docs.gitlab.com/ee/user/packages/package_registry/index.html) and can be used in combination with [regex managers](https://docs.renovatebot.com/modules/manager/regex/) to keep dependencies up-to-date which are not specifically supported by Renovate.

> To specify which specific repository should be queried when looking up a package, the `registryUrl` has to be set like this: `https://gitlab.com/user/project`.

Reviewer: ```suggestion
To specify which specific repository should be queried when looking up a package, the `registryUrl` has to be set following this format: `https://gitlab.com/user/project`.
```

I think we're trying to explain that you should follow a specific format when setting the `registryUrl`.

As an example, `https://gitlab.com/gitlab-org/ci-cd/package-stage/feature-testing/new-packages-list` would look for packages in the generic packages repository of the `gitlab-org/ci-cd/package-stage/feature-testing/new-packages-list` project.

If you are using a self-hosted GitLab instance, please note the following requirements:

- If you are on the `Free` edition, this datasource requires at least GitLab 13.3

> - If you are on the `Premium` or the `Ultimate` edition, this datasource requires at least GitLab 11.8, but GitLab 12.9 or more is recommended if you have a lot of packages with different name on the same project.

Reviewer: ```suggestion
- If you are on the `Premium` or the `Ultimate` edition, this datasource requires at least GitLab 11.8, but GitLab 12.9 or more is recommended if you have a lot of packages with different name on the same project
```

We don't use a full stop `.` at the end of list items.

> - If you are on the `Premium` or the `Ultimate` edition, this datasource requires at least GitLab 11.8, but GitLab 12.9 or more is recommended if you have a lot of packages with different name on the same project

Reviewer: ```suggestion
- If you are on the `Premium` or the `Ultimate` edition, this datasource requires at least GitLab 11.8, but GitLab 12.9 or more is recommended if you have a lot of packages with different names on the same project
```

I think we should use the plural form here?

**Usage Example**

> A real-world example for this specific datasource would be maintaining packages versions in a config file.

Reviewer: ```suggestion
A real-world example for this specific datasource would be maintaining package versions in a config file.
```

I think we can use the singular form here.
